// react-relay-network-modern/es/express-middleware/graphqlBatchHTTPWrapper.js
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return typeof key === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (typeof input !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (typeof res !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
export default function (graphqlHTTPMiddleware) {
  return (req, res) => {
    const subResponses = [];
    return Promise.all(req.body.map(data => new Promise(resolve => {
      const subRequest = _objectSpread(_objectSpread({
        __proto__: req.__proto__
      }, req), {}, {
        body: data
      });
      const subResponse = _objectSpread(_objectSpread({}, res), {}, {
        status(st) {
          this.statusCode = st;
          return this;
        },
        set() {
          return this;
        },
        send(payload) {
          resolve({
            status: this.statusCode,
            id: data.id,
            payload
          });
        },
        // support express-graphql@0.5.2
        setHeader() {
          return this;
        },
        header() {},
        write(payload) {
          this.payload = payload;
        },
        end(payload) {
          // support express-graphql@0.5.4
          if (payload) {
            this.payload = payload;
          }
          resolve({
            status: this.statusCode,
            id: data.id,
            payload: this.payload
          });
        }
      });
      subResponses.push(subResponse);
      graphqlHTTPMiddleware(subRequest, subResponse);
    }))).then(responses => {
      let response = '';
      responses.forEach(({
        status,
        id,
        payload
      }, idx) => {
        if (status) {
          res.status(status);
        }
        const comma = responses.length - 1 > idx ? ',' : '';
        response += `{ "id":"${id}", "payload":${payload} }${comma}`;
      });
      res.set('Content-Type', 'application/json');
      res.send(`[${response}]`);
    }).catch(err => {
      res.status(500).send({
        error: err.message
      });
    });
  };
}
